Use Windows matcher when on Windows platform in all code paths

jsturtevant opened the pull request:
When looking into https://github.com/containerd/nerdctl/issues/751 I found that the platform Normalization function is dropping the `OSVersion` because it was marked as depreciated but this value is used throughout the windows https://github.com/containerd/containerd/blob/9c676e98dde995db46841a2c6126ad709e3fc7d7/platforms/defaults_windows.go#L36

There maybe more going on here than just this change.  For example, this may be why this code uses `Ordered` here and then adds the `osVersionPrefix`. 


This change does make it so that nerdctl will work with multiarch manifests but there may need to be more work done here.

AkihiroSuda commented:
/ok-to-test

jsturtevant commented:
/hold 
I was testing some more and this may not solve it fully

kevpar commented:
This looks good based on the previous investigation I did: https://github.com/containerd/containerd/issues/6168#issuecomment-991313576

kzys reviewed (changes requested):
Can you add some tests? Windows' platform handling seems really tricky...

jsturtevant commented:
> This looks good based on the previous investigation I did: [#6168 (comment)](https://github.com/containerd/containerd/issues/6168#issuecomment-991313576). Although we probably don't want to unset `OSFeatures` either.

There is a bit more going on. I have a partial fix that I can push up but it caused some other issues around LOCW support on windows.  The [LCOW support](https://github.com/containerd/containerd/blob/c2cb589221de30eb12882a2a83b6f8441209352a/platforms/defaults_windows.go#L85-L87) doesn't really work since the matcher would never pass the [osPrefix version check](https://github.com/containerd/containerd/blob/c2cb589221de30eb12882a2a83b6f8441209352a/platforms/defaults_windows.go#L50-L52)

The main thing is on Windows some code paths (such as [image pulling](https://github.com/containerd/containerd/blob/c2cb589221de30eb12882a2a83b6f8441209352a/pull.go#L56)) are using the [windows matcher](https://github.com/containerd/containerd/blob/c2cb589221de30eb12882a2a83b6f8441209352a/platforms/defaults_windows.go#L49-L55) and other it is using [linux matcher](https://github.com/containerd/containerd/blob/c2cb589221de30eb12882a2a83b6f8441209352a/platforms/platforms.go#L148-L153).   This is because the [Ordered function news up a matcher](https://github.com/containerd/containerd/blob/c2cb589221de30eb12882a2a83b6f8441209352a/platforms/compare.go#L112-L115) that will be [Linux default matcher](https://github.com/containerd/containerd/blob/c2cb589221de30eb12882a2a83b6f8441209352a/platforms/platforms.go#L138-L142) that doesn't include OSVersion Prefix matching. 

In the case of Windows image pull scenario with ctr this means it matched and pulled both Windows version as described in https://github.com/containerd/nerdctl/issues/751 (comment: `If it helps, I noticed something weird when I run nerdctl.exe images. When I run the command, it prints the container image twice for some reason:`)

jsturtevant commented:
I added https://github.com/containerd/containerd/pull/6491/commits/5ae90c60c7c5ca5a30e282426dd49add7e8a665f commit to demonstrate my comments in https://github.com/containerd/containerd/pull/6491#issuecomment-1026170393

In particular that:

- The LCOW support doesn't really work since the matcher would never pass the osPrefix version check
- This is because the Ordered (and only) function news up a matcher that will be Linux default matcher that doesn't include OSVersion Prefix matching.

thaJeztah commented on platforms/compare_test.go:
```diff
@@ -75,6 +81,18 @@ func TestOnly(t *testing.T) {
 			},
 		},
+		{
+			platform: fmt.Sprintf("windows/amd64//%s", buildStr+".1"),
```
Slightly orthogonal, but wondering if the string representation for osversion (and features) was already "formalised" . I recall there were some discussion about this in the past (but don't recall the outcome). We should probably have some of this added to the OCI specs at some point

jsturtevant replied:
I wasn't sure if this valid string.  The parser doesn't actually work in this case but was trying to demonstrate this isn't really handled.

thaJeztah replied:
Makes sense! Yes, I also was looking at Parse() not yet supporting it (and left that for a later exercise).

Empty `//` may work for missing components (but could potentially be problematic if someone (eg) uses `path.Join()` to construct the string).

Perhaps there's alternatives (`/-/` for explicitly omitting a field, or `/*/`); in any case, likely something for further discussion

thaJeztah replied:
Discussing with someone else, and an alternative for the format could be `os/arch/variant@os-version[+feature+feature...]`

(we should look at the `+` symbol though, in case  it may be used in URLs (where `+` can be ambiguous)

jsturtevant commented:
> Would LCOW use the _default_ matcher though? Given that in the LCOW case, the container effectively runs "remotely", wouldn't it require a matcher that's created based on that environment?

This might be the case but I not sure every architecture/variant of Linux will run on Windows so we may need a custom one?  Maybe @kevpar knows?

jsturtevant commented:
I've updated to use the appropriate matcher on a given OS. On Windows it will use a Windows Matcher which uses the default matcher and only adds the additional checks if the platform specific is Windows.  Otherwise, the behavior should be the same which is to use the default matcher

jsturtevant commented:
This last commit works with the original issue reported (containerd/nerdctl#751 )

```
PS C:\> .\nerdctl.exe run --rm --net none hello-world:latest
docker.io/library/hello-world:latest:                                             resolved       |++++++++++++++++++++++++++++++++++++++|
index-sha256:507ecde44b8eb741278274653120c2bf793b174c06ff4eaa672b713b3263477b:    exists         |++++++++++++++++++++++++++++++++++++++|
manifest-sha256:b17da88a43678dd8de2c335b977330a5f4e3f1175251a4204f54ef2ed1360709: exists         |++++++++++++++++++++++++++++++++++++++|
config-sha256:bbca594eee38d2bcbda18490d2667560b2452e74e73b8c26bddea59cb6920ad1:   done           |++++++++++++++++++++++++++++++++++++++|
elapsed: 0.9 s                                                                    total:   0.0 B (0.0 B/s)

Hello from Docker!
This message shows that your installation appears to be working correctly.

To generate this message, Docker took the following steps:
 1. The Docker client contacted the Docker daemon.
 2. The Docker daemon pulled the "hello-world" image from the Docker Hub.
    (windows-amd64, nanoserver-1809)
 3. The Docker daemon created a new container from that image which runs the
    executable that produces the output you are currently reading.
 4. The Docker daemon streamed that output to the Docker client, which sent it
    to your terminal.

To try something more ambitious, you can run a Windows Server container with:
 PS C:\> docker run -it mcr.microsoft.com/windows/servercore:ltsc2019 powershell

Share images, automate workflows, and more with a free Docker ID:
 https://hub.docker.com/

For more examples and ideas, visit:
 https://docs.docker.com/get-started/

PS C:\> cmd /c ver

Microsoft Windows [Version 10.0.17763.2452]
PS C:\> .\nerdctl.exe --version
nerdctl version 0.16.0-39-g2f0cf75.m
```

gabriel-samfira commented on platforms/defaults_windows.go:
```diff
@@ -78,14 +81,21 @@ func revision(v string) int {
+	if len(parts) < 4 {
+		return v
+	}
+	return strings.Join(parts[0:2], ".")
```
This will only return the first 2 elements of parts, if the length is >= 4. Which means you'll match ```10.0.20348``` as well as ```10.0.17763```. I _think_ this should be ```return strings.Join(parts[0:3], ".")```

jterry75 replied:
Why cant we just match the build number. That is all that is required for process isolation compat right? There is never going to be a 10.0.17763 and a 11.0.17763 :).

gabriel-samfira replied:
Sure, that also works. If you believe that the build number will always be unique, whatever works to match the image in the manifest to the version of windows that's running. However, I believe that elsewhere in the code, the major, minor and build numbers are used to compare versions wherever needed.

I pointed out the missing build nr here, because a prefix generated by this would match both Windows Server 2019 and Windows Server 2022. (```10.0```)

jterry75 replied:
Sorry @gabriel-samfira for the confusion - You are 100% correct. I agree with you if we do not just use build numbers.

I was asking @jsturtevant if we should just use build numbers since that is the kernel compat level. But its likely just safer to do the major.minor.build

thaJeztah commented on platforms/defaults_windows.go:
```diff
@@ -78,14 +81,21 @@ func revision(v string) int {
-			Architecture: runtime.GOARCH,
-		}),
-		osVersionPrefix: fmt.Sprintf("%d.%d.%d", major, minor, build),
+	// Windows also supports linux via LCOW
+	linux := specs.Platform{
+		OS: "linux",
+		Architecture: "amd64",
 	}
```
I'm not sure if this applies everywhere. I think the default matcher on Windows should remain to be "Windows" (and vice-versa for Linux), but it should be possible for consumers of the containerd code (like nerdctl) to easily construct a matcher for other platforms

I guess this would already be possible if `nerdctl` on Windows (if LCOW is available) to use `platform.Ordered()` or `platform.Any()` https://github.com/containerd/containerd/blob/a43703fcba5432751059575bbd89d578aed69101/platforms/compare.go#L110-L120

Perhaps a utility function that allows to combine multiple matchers, something like;
 
```go
func NewOrderedPlatformComparer(matchers ...Matcher) MatchComparer {
	return orderedPlatformComparer{
		matchers: matchers,
	}
}
```

jterry75 replied:
Interesting idea. Basically you would turn off LCOW for anybody that doesnt explicitly turn it on? So in theory we would need to do this in Docker, CRI, nerdctl, ctr, and anyone else that calls containerd directly

thaJeztah replied:
Yes, so I'm not sure if LCOW will be available on every Windows set up (even if that's the case, I think the _default_ match should likely be best matching "windows" (native) image).

That said; containerd is meant to be used as component to build a "final" product/project, and from that perspective should be less opinionated (but flexible), so I think it would make sense to allow the project/code that uses it to construct an appropriate matcher for their purpose.

jterry75 replied:
I'm on-board even though I don't get a vote lol. LCOW is not possible for everyone to use and so pulling a LCOW layer of a multi-os image seems wasteful by default on Windows.

jsturtevant replied:
>I guess this would already be possible if nerdctl on Windows (if LCOW is available) to use platform.Ordered() or platform.Any()

It is possible and why I found this issue with the matchers because it was using those functions which used the "default" matcher which doesn't match windows with the prefixes.

It makes sense to me to drop this and allow it to be passed through were appropriate. The tests now cover the scenario where a user might want to create their own LCOW matcher on Windows.

For why this is here, I carried this through from the current implementation (though it [never matched properly](https://github.com/containerd/containerd/blob/604c462d7bc59f967511c923e21c86014a4d0891/platforms/defaults_windows.go#L50-L52)): https://github.com/containerd/containerd/blob/604c462d7bc59f967511c923e21c86014a4d0891/platforms/defaults_windows.go#L84-L88

I also saw this on Darwin matcher: https://github.com/containerd/containerd/blob/604c462d7bc59f967511c923e21c86014a4d0891/platforms/defaults_darwin.go#L40-L43

jsturtevant replied:
removed linux from the default matcher on windows

gabriel-samfira commented on platforms/platforms.go:
```diff
@@ -261,6 +261,5 @@ func Normalize(platform specs.Platform) specs.Platform {
 	// these fields are deprecated, remove them
 	platform.OSFeatures = nil
 	platform.OSVersion = ""
```
This also needs to be removed. Otherwise, when we normalize the platform, the OSVersion gets dropped and the matcher will match all versions of Windows from a manifest (1809, ltsc2022, etc).

jsturtevant replied:
thanks! I had this initially and dropped it during my updates. I can add it back here or rebase after  #6497

dmcgowan commented on platforms/platforms.go:
```diff
@@ -127,15 +127,15 @@ type Matcher interface {
 	Match(platform specs.Platform) bool
 }
 
-// NewMatcher returns a simple matcher based on the provided platform
```
Best to have the exported function in the non-platform specific file then call into non-exported implementations. This will also need an other implementation besides just Linux and Windows

jterry75 replied:
+1

jsturtevant replied:
> This will also need an other implementation besides just Linux and Windows

I've updated to include other implementations, it uses the same defaults as previous except Windows.

> Best to have the exported function in the non-platform specific file then call into non-exported implementations

I don't think I fully understood what I should do here.  could you explain a bit more? Thanks!

jterry75 replied:
`NewMatcher` is an exported function. So it should be in platforms.go which calls the un-exported `newDefaultMatcher` in the various platforms_<os>.go files. But actually thats another point. Why is `NewDefaultMatcher` exported? There are no callers right?

jsturtevant replied:
Thanks, I see what I did and I've updated

jterry75 commented on platforms/defaults_windows.go:
```diff
@@ -78,14 +81,21 @@ func revision(v string) int {
+	// Windows also supports linux via LCOW
+	linux := specs.Platform{
+		OS: "linux",
+		Architecture: "amd64",
```
LCOW wont match x86?

jsturtevant replied:
woops I should have had `runtime.GOARCH` https://github.com/containerd/containerd/blob/604c462d7bc59f967511c923e21c86014a4d0891/platforms/defaults_windows.go#L87

based on https://github.com/containerd/containerd/blob/604c462d7bc59f967511c923e21c86014a4d0891/platforms/defaults_windows.go#L87 sounds like we will remove this anyways

jterry75 commented on platforms/defaults_windows_test.go:
```diff
@@ -47,15 +47,39 @@ func TestDefault(t *testing.T) {
+		{
+			platform: imagespec.Platform{
+				OS: "linux",
+				Architecture: runtime.GOARCH,
```
This would fail if ever run on x86 machine. Do we care?

jsturtevant replied:
this should be updated if we dont' remove it based on https://github.com/containerd/containerd/pull/6491/files#r797781340

TBBle commented on platforms/defaults_windows.go:
```diff
@@ -39,25 +39,28 @@ func DefaultSpec() specs.Platform {
+	match := m.defaultMatcher.Match(p)
+
+	if match && m.OS == "windows"  {
+		return strings.HasPrefix(p.OSVersion, m.osVersionPrefix) && m.defaultMatcher.Match(p)
```
Windows Server 2022 has made this more complex, as its kernel (`10.0.20348.x`) is now a stable ABI, and so Windows hosts with greater kernel builds (only currently Windows 11, `10.0.22000.y), can run container images marked `10.0.20348.z` in process isolation mode, and hence should pass this test.

More details at https://github.com/microsoft/Windows-Containers/issues/117#issuecomment-912645956, https://techcommunity.microsoft.com/t5/containers/windows-server-2022-and-beyond-for-containers/ba-p/2712487, and https://github.com/Microsoft/hcsshim/blob/master/test/cri-containerd/main_test.go#L162-L212.

And also https://github.com/microsoft/Windows-Containers/issues/170#issuecomment-1024520322, where I recently discovered the issue being addressed here and in #6497, and forgot to raise a bug.

(On that note, if we want to punt the down-level support into a new containerd issue, that's fine too.)

jsturtevant replied:
my vote is to handle this in separate issue.

TBBle replied:
Done as #6508.

jsturtevant commented:
Thanks for the feedback everyone! I will address it shortly and squash the commits.

TBBle commented on platforms/defaults_windows.go:
```diff
@@ -39,25 +39,28 @@ func DefaultSpec() specs.Platform {
-	if m.defaults.Match(p) {
-		// TODO(windows): Figure out whether OSVersion is deprecated.
-		return strings.HasPrefix(p.OSVersion, m.osVersionPrefix)
+func (m windowsmatcher) Match(p imagespec.Platform) bool {
```
I realise this isn't part of this change, but is there any reason to have `imagespec` _and_ `specs` in this file? They appear to reference the same module, and the only uses of `imagespec` are both touched by this commit, so maybe now's the time to replace the two uses of `imagespec` with `specs`.

Or perhaps there's actually a distinction I'm overlooking.

jsturtevant replied:
I don't think you're overlooking anything, and I've updated to use specs with is used in the other files

jsturtevant commented:
I've addressed all the feedback and this is ready for review.  

The node test failure was a Prow pod timeout.  not sure if the same commands work here:
/retest

thaJeztah commented:
One thing I just realised; so this changes the default on windows to match on architecture _and_ os-version, correct? I think that's correct for containers running with process isolation, but would this be an issue when running with hyper-v isolation? (not sure of that's supported by containerd, but I know dockerd allows switching)

jterry75 commented:
@kevpar - Is there such a thing as x86 hypervisor isolated on Windows? I've never seen such an image to my knowledge. Would Seb's comment actually be possible?

jterry75 commented:
Yea, ..., no good answer here imo. I think I agree with @thaJeztah earlier comment. Windows should match Windows process execution by default. Only the caller has enough context to know the intent of the container that will be scheduled next. IE: The caller knows if the container config is isolated or not. But it does not tell that to containerd when pulling images. But it CAN by providing a different matcher when pulling images. So I think the right default is Windows process works as expected, and if hypervisor is needed caller can control it.

jterry75 commented:
@jsturtevant - Any chance you have seen this?

```
PS C:\Users\Administrator\Desktop\Tools> ctr image ls
time="2022-03-01T18:40:26Z" level=error msg="failed calculating size for image mcr.microsoft.com/windows/servercore:10.0.17763.2458" error="no match for platform in manifest: not found"
time="2022-03-01T18:40:26Z" level=error msg="failed calculating size for image mcr.microsoft.com/windows/servercore:10.0.17763.2565" error="no match for platform in manifest: not found"
time="2022-03-01T18:40:26Z" level=error msg="failed calculating size for image mcr.microsoft.com/windows/servercore:ltsc2019" error="no match for platform in manifest: not found"
REF                                                       TYPE                                                      DIGEST                                                                  SIZE    PLATFORMS     LABELS
mcr.microsoft.com/windows/servercore:10.0.17763.2458      application/vnd.docker.distribution.manifest.list.v2+json sha256:7d370baf7e097a20bdd52d5a214fcad25f83caccca049cced5d5eb64d659c45e 480.0 B windows/amd64 -
mcr.microsoft.com/windows/servercore:10.0.17763.2565      application/vnd.docker.distribution.manifest.list.v2+json sha256:08968b1a61ea651efa830ade84c332df560bf5b551a94bb065c128989deb346d 480.0 B windows/amd64 -
mcr.microsoft.com/windows/servercore:1809-KB5010791-amd64 application/vnd.docker.distribution.manifest.v2+json      sha256:80b95537f26ad6f44e9d3675a0f4be3a076364dd1f988b7f0ceb3190eb31115d 2.5 GiB windows/amd64 -
mcr.microsoft.com/windows/servercore:ltsc2019             application/vnd.docker.distribution.manifest.list.v2+json sha256:d8c2b69511a49e9eb6fd2a17f2727ae011b0382cd11562fa5cb4c70bb460cf12 480.0 B windows/amd64 -
```
Seems related to the change. But I can look closer if you are unaware. 

Oddly, the images pull just fine, and unpack just fine, and run just fine.

jsturtevant commented:
I am not, my suspicion is the matcher or string to match with is wired up wrong for that code path in ctr.  This looks similar to what I was seeing nerdctl when I turned on debug mode.

As a point of reference, We have a nightly job for containerd which has been running well since this change merged https://testgrid.k8s.io/sig-windows-master-release#aks-engine-windows-containerd-nightly

jterry75 commented:
Oh, I see. It looks like we have broken the unpacker for platform agnostic variants when the Host os doesn't match (meaning won't unpack so Xenon cant even run). So for example:

On WS2022:

```
ctr image pull mcr.microsoft.com/windows/servercore:ltsc2019
ctr image pull mcr.microsoft.com/windows/servercore:ltsc2019
mcr.microsoft.com/windows/servercore:ltsc2019:                                    resolved       |++++++++++++++++++++++++++++++++++++++|
index-sha256:d8c2b69511a49e9eb6fd2a17f2727ae011b0382cd11562fa5cb4c70bb460cf12:    done           |++++++++++++++++++++++++++++++++++++++|
manifest-sha256:596c8a02a8e73b53688052907d121879357d2a9208700306bafaa1ffb533cbab: done           |++++++++++++++++++++++++++++++++++++++|
layer-sha256:1bd78008c728d8f9e56dc2093e6eb55f0f0b1aa96e5d0c7ccc830c5f60876cdf:    exists         |++++++++++++++++++++++++++++++++++++++|
config-sha256:c703d4d605332bcb954f514c7d2c6cd53ee51024d7a7a3f13eb907abbeedbca2:   done           |++++++++++++++++++++++++++++++++++++++|
layer-sha256:4612f6d0b889cad0ed0292fae3a0b0c8a9e49aff6dea8eb049b2386d9b07986f:    exists         |++++++++++++++++++++++++++++++++++++++|
elapsed: 0.6 s                                                                    total:  1.3 Ki (2.2 KiB/s)
unpacking windows/amd64 sha256:d8c2b69511a49e9eb6fd2a17f2727ae011b0382cd11562fa5cb4c70bb460cf12...
ctr: no match for platform in manifest sha256:d8c2b69511a49e9eb6fd2a17f2727ae011b0382cd11562fa5cb4c70bb460cf12: not found
```

But if you add the `-amd64`
```
ctr image pull mcr.microsoft.com/windows/servercore:ltsc2019-amd64
mcr.microsoft.com/windows/servercore:ltsc2019-amd64:                              resolved       |++++++++++++++++++++++++++++++++++++++|
manifest-sha256:596c8a02a8e73b53688052907d121879357d2a9208700306bafaa1ffb533cbab: exists         |++++++++++++++++++++++++++++++++++++++|
config-sha256:c703d4d605332bcb954f514c7d2c6cd53ee51024d7a7a3f13eb907abbeedbca2:   exists         |++++++++++++++++++++++++++++++++++++++|
layer-sha256:1bd78008c728d8f9e56dc2093e6eb55f0f0b1aa96e5d0c7ccc830c5f60876cdf:    exists         |++++++++++++++++++++++++++++++++++++++|
layer-sha256:4612f6d0b889cad0ed0292fae3a0b0c8a9e49aff6dea8eb049b2386d9b07986f:    exists         |++++++++++++++++++++++++++++++++++++++|
elapsed: 0.2 s                                                                    total:   0.0 B (0.0 B/s)
unpacking windows/amd64 sha256:596c8a02a8e73b53688052907d121879357d2a9208700306bafaa1ffb533cbab...
done: 3m12.8486039s
```

The `mcr.microsoft.com/windows/servercore:ltsc2019-amd64` is actually a manifest list which I why I think it breaks for only that variant.

jterry75 commented:
@TBBle - I've already forgotten :(. Did we decide it was "ok" to break hv isolated pulls when the host os doesnt match the image hahaha? And that to do this would require the caller to pass in a custom matcher? I think my brain is giving out

jsturtevant commented:
https://github.com/containerd/containerd/pull/6491#issuecomment-1036828304 😄 

The fact that it worked before was accidental IMO.  I am still not sure if hyper-v upstream is [fully functional](https://github.com/containerd/containerd/pull/6491#issuecomment-1041584233), talking with @dcantah it sounds like it isn't

kevpar commented:
My $0.02:

I think for cases like nerdctl and ctr we should probably have separate flag(s) to control the exact matching behavior so that cases like hypervisor-isolation can work. In CRI my view is we should add fields to [`Runtime`](https://github.com/containerd/containerd/blob/main/pkg/cri/config/config.go#L32) to control image matching behavior. We may need to extend CRI's `PullImage` to take a runtime name, but that would allow pulling an image for a specific runtime, and the runtime controls what image variant is pulled.

One difficulty is last I checked containerd doesn't have great support for multi-arch images. I think there were places where it's assumed we're only working with a single platform at a time.

jterry75 commented:
> 

So funny.

jterry75 commented:
> [#6491 (comment)](https://github.com/containerd/containerd/pull/6491#issuecomment-1036828304) 😄
> 
> The fact that it worked before was accidental IMO. I am still not sure if hyper-v upstream is [fully functional](https://github.com/containerd/containerd/pull/6491#issuecomment-1041584233), talking with @dcantah it sounds like it isn't

So funny.

jterry75 commented:
@kevpar - Agreed. I think when I get around to it I will add something like:
`ctr image pull --isolated <>` or something tricky like that to indicate that I'm ok with some set of partial matches that honor how hypervisor compat is supposed to work.

TBBle commented:
Yeah, that's what I took away from the conversation as well: Take Linux out of the containerd default Windows-host matcher, and let tools that are happy to support Hyper-V isolation and/or WCOW (i.e. Xenon aka UVM) provide a matcher to suit.

It might be feasible to add a function that creates a Xenon-capable matcher, to save some code duplication in clients, and hence centralise any particular complex or specific rules that might appear. It might also make sense to put something like that in hcsshim or other central-but-Windows-host-specific repo instead.

But that assumes a tool that wants to integrate Xenon wants to accept both kinds of matches, so maybe it's not _that_ valuable after all.

For CRI, I noticed that `ImageSpec` given to `PullImage` specifically [describes its `Annotation` field for almost this use-case](https://github.com/kubernetes/cri-api/blob/master/pkg/apis/runtime/v1/api.proto#L670-L673):
```
    // Unstructured key-value map holding arbitrary metadata.
    // ImageSpec Annotations can be used to help the runtime target specific
    // images in multi-arch images.
    map<string, string> annotations = 2;
```

It does make sense that if you are using `Runtime` to choose Isolation mode and other config options, then that should be visible to `PullImge` too; the backend can ignore the Runtime in the case where it doesn't matter. It could probably be plumbed through the Annotations pending the next spec update.

jterry75 commented:
@TBBle - That's exactly why we added `annotations` to the CRI Image :). Glad it's finally paying off.

I think we have consensus haha. We are good to go with the code as is.
